File: book2-algorithms-sedgewick/ch2/06_quick_sort.py
from math import floor
import sys
import random
import time
from typing import Any, Sequence

class QuickSort:

    def sort(self, items: Sequence[Any]) -> None:
        self._sort(items, 0, len(items) - 1)
    
    def _sort(self, items: Sequence[Any], left: int, right: int) -> None:
        if left >= right: return
        if left < 0 or right >= len(items): return
        j = self._partition(items, left, right)
        self._sort(items, left, j - 1)
        self._sort(items, j + 1, right)

    def _partition(self, items: Sequence[Any], left: int, right: int) -> None:
        i = left
        j = right + 1
        v = items[left]
        while True:
            i += 1
            while items[i] < v:
                if i >= right: break
                i += 1
            j -= 1
            while items[j] > v:
                if j <= left: break
                j -= 1
            if i >= j: break
            items[i], items[j] = items[j], items[i]
        items[left], items[j] = items[j], items[left]
        return j


def main():
    print('QuickSort.')

    if len(sys.argv) < 2:
        print('Missing number of items.')
        exit(1)
    
    number_items = int(sys.argv[1])
    print('Number of Items: ', number_items)

    print('Generating numbers... ', end='', flush=True)
    numbers = [random.randint(0, 1000000) for _ in range(number_items)]
    print('Done.')

    if number_items <= 100:
        print('Numbers: ', ', '.join((str(n) for n in numbers)))
    
    print('Sorting... ', end='', flush=True)
    quick_sort = QuickSort()
    t1 = time.time()
    quick_sort.sort(numbers)
    t2 = time.time()
    print('Done.')

    print('Running time: ', (t2 - t1), 'secs.')

    if number_items <= 100:
        print('Numbers: ', ', '.join((str(n) for n in numbers)))

    print('Checking... ', end='', flush=True)
    for i in range(len(numbers) - 1):
        if numbers[i] > numbers[i + 1]:
            print('\nBad ordering: ', numbers[i], ', ', numbers[i + 1])
            exit(1)
    print('Done.')

if __name__ == '__main__':
    main()

File: book2-algorithms-sedgewick/ch2/06_quick_sort.cpp
#include <iostream>
#include <algorithm>
#include <utility>
#include <random>
#include <string>
#include <vector>
#include <chrono>

template<typename T>
class QuickSort {
public:

    QuickSort() {};

    void sort(T& arr) {
        sort(arr, 0, arr.size() - 1);
    }

private:

    void sort(T& arr, size_t left, size_t right) {
        if (left >= right) return;
        if (left < 0 || right >= arr.size()) return;
        size_t j = partition(arr, left, right);
        sort(arr, left, j - 1);
        sort(arr, j + 1, right);
    }

    size_t partition(T& arr, size_t left, size_t right) {
        auto i = left;
        auto j = right + 1;
        auto v = arr[left];
        while (true) {
            while (arr[++i] < v) if (i >= right) break;
            while (v < arr[--j]) if (j <= left) break;
            if (i >= j) break;
            swap(arr, i, j);
        }
        swap(arr, left, j);
        return j;
    }

    void swap(T& arr, size_t i, size_t j) {
        auto tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
};

int main(int argc, char* argv[]) {
    using std::chrono::high_resolution_clock;
    using std::chrono::duration_cast;
    using std::chrono::milliseconds;

    std::cout << "Quicksort." << std::endl;

    if (argc <= 1) {
        std::cerr << "Missing number of elements." << std::endl;
        return EXIT_FAILURE;
    }

    size_t numbers_length = std::stoi(argv[1]);
    std::cout << "Number of Elements: " << numbers_length << std::endl;

    std::cout << "Generating data... " << std::flush;
    std::vector<long> numbers;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution <size_t> dist(0, 1000000);
    while (numbers_length > 0) {
        numbers.push_back(dist(gen));
        numbers_length--;
    }
    std::cout << "Done." << std::endl;

    if (numbers.size() < 100) {
        auto it = numbers.begin();
        std::cout << "Numbers: ";
        while (it != numbers.end()) {
            std::cout << *it << " ";
            it++;
        }
        std::cout << std::endl;
    }

    std::cout << "Sorting... " << std::flush;
    auto t1 = high_resolution_clock::now();
    QuickSort<std::vector<long>> quick_sort;
    quick_sort.sort(numbers);
    auto t2 = high_resolution_clock::now();
    std::cout << "Done." << std::endl;
    auto running_time = duration_cast<milliseconds>(t2 - t1).count();
    std::cout << "Running Time: " << running_time << "ms" << std::endl;

    if (numbers.size() < 100) {
        auto it = numbers.begin();
        std::cout << "Numbers: ";
        while (it != numbers.end()) {
            std::cout << *it << " ";
            it++;
        }
        std::cout << std::endl;
    }

    std::cout << "Checking... " << std::flush;
    for (size_t i = 0; i < numbers.size() - 1; i++) {
        if (numbers[i] > numbers[i + 1]) {
            std::cerr << "Bad sorting: " << numbers[i] << ", " << numbers[i + 1] << std::endl;
        }
    }
    std::cout << "Done." << std::endl;

    return EXIT_SUCCESS;
}